src/drivers/imx/interrupt.c

```diff
@@ -3,26 +3,443 @@
+ * ---> IRQ 0-63 ---> |   |          |    |           |   |  |
+ * |  [MASK 0-63] --> | & | --/64--> | OR | ---/1---> | & | ----> OUT[0]
+ * |                  |   | [STATUS] |    | [MD0] --> |   |  |
+ * |                  +---+          +----+           +---+  |
+ * |                                                         |
+ * | ... (same for the other IRQ lines and outputs to OUT[7] |
+ * |                                                         |
+ * +---------------------------------------------------------+
+ *
+ * In the schematic above:
+ *    IRQ 0-511: Input IRQ lines (shared IRQs). IRQs 0-31 are reserved.
+ *    MASK 0-511: Configurable mask for interrupts.
+ *    MD0-MD7: Master disable register, block an entire output interrupt
+ *    line.
+ *    STATUS: Read only register which shows what interrupts are active.
+ *    OUT: The 8 interrupt lines that lead to the DSP, leading to arch
+ *    IRQs IRQ_NUM_IRQSTR_DSP0 through 7.
+ *
+ * Usage of the hardware: We turn on the hardware itself, then we
+ * configure the mask (all mask bits default to 0), enable our arch
+ * interrupts and wait for an interrupt on an output line.
+ *
+ * Upon receiving an arch interrupt, the driver must check the STATUS
+ * registers corresponding to the arch interrupt in order to figure out
+ * what the actual, input shared interrupt was, and then call any
+ * registered callback to handle the condition leading to the interrupt.
+ *
+ * The hardware also supports forcing an interrupt from the software; I
+ * have omitted this from the schematic since it is not relevant to the
+ * usage in this driver.
+ */
+
+#define IRQSTR_BASE_ADDR_QXP	0x51080000
+#define IRQSTR_BASE_ADDR	IRQSTR_BASE_ADDR_QXP
+
+/* The MASK, SET (unused) and STATUS registers are 512-bit registers
+ * split into 16 32-bit registers that we can directly access.
+ *
+ * To get the proper register for the shared interrupt irq, we do
+ * IRQSTR_CH_MASK(IRQSTR_INT_REG(irq)) (MASK can be replaced by SET or
+ * STATUS).
+ *
+ * The interrupts are mapped in the registers in the following way:
+ * Interrupts 480-511 at offset 0
+ * Interrupts 448-479 at offset 1
+ * Interrupts 416-447 at offset 2
+ * ...
+ * Interrupts 64-95 at offset 13
+ * Interrupts 32-63 at offset 14
+ * Interrupts 0-31 at offset 15
+ *
+ * The IRQSTR_CH_* macros perform the second part of this calculation
+ * (offset) automatically.
+ */
+
+#define IRQSTR_CHANCTL		0x00
+#define IRQSTR_CH_MASK(n)	(0x04 + 0x04 * (15 - (n)))
+#define IRQSTR_CH_SET(n)	(0x44 + 0x04 * (15 - (n)))
+#define IRQSTR_CH_STATUS(n)	(0x84 + 0x04 * (15 - (n)))
+#define IRQSTR_MASTER_DISABLE	0xC4
+#define IRQSTR_MASTER_STATUS	0xC8
+
+#define IRQSTR_INT_REG(irq)	((irq) / 32)
+#define IRQSTR_INT_BIT(irq)	((irq) % 32)
+#define IRQSTR_INT_MASK(irq)	(1 << IRQSTR_INT_BIT(irq))
+
+#define IRQSTR_RESERVED_IRQS	32
+#define IRQSTR_IRQS		512
+#define NUM_IRQ_REGISTERS	16 /* Number of IRQ_CH_MASK/SET/STATUS regs */
+
```

Comment:
We should keep IRQSTR_ prefix to all macros. Better names for macros above:

IRQSTR_RESERVED_IRQS_NUM
IRQSTR_IRQS_NUM
IRQSTR_IRQS_REGISTERS_NUM